### DISCUSSION TITLE
Open Source

### DISCUSSION THREAD

**Discussion by igor-lemon** [DockStation/dockstation#311] [Q&A]
**Is the project Open Source?**
- No

**Will you publish the source code?**
- No, I don't have such plans

**Why you don't publish source code?**
- There are many reasons for this, but let's just assume that I can't to do yet.

**I want the source code!**
- If it's important for you, just use Open Source alternatives:
https://medium.com/@dockstation/a-brief-comparison-of-gui-docker-tools-46bd6a24ae31

**Why do you have GitHub repo?**
- Because GitHub has very convenient ticket system for issues discussing.

**Comment by 0xdeada55** [comment]
> * There are many reasons for this, but let's just assume that I don't want to do yet.

That sounds, sketchy. Anyway, I wish you luck with the project.

**Reply by igor-lemon** [reply]
- I'm not the only one copyright holder and I can't do anything without approve of all owners
- Not the best development practices applied inside because it was us first project which we made on React without any mentoring. 
Rewrite the app from scratch will be faster than refactor it, I'm 100% sure no one will do it.
- If you afraid about code just discover it. The app based on Electron, unpack `asar` archive then check and discover the code.

**Reply by 0xdeada55** [reply]
> * If you afraid about code just discover it. The app based on Electron, unpack `asar` archive then check and discover the code.

Don't you think I already did that? It wasn't of much help when everything (renderer+main code) is highly minified/compressed and node_modules are inlined in a single file. I wonder what is the reason for that

**Reply by igor-lemon** [reply]
>Don't you think I already did that? It wasn't of much help when everything (renderer+main code) is highly minified/compressed and node_modules are inlined in a single file. I wonder what is the reason for that

The reason is that I use webpack for make bundles.
Very very weak minification. I don’t even do improved minification, much less code obfuscation, although it could.

**Reply by 0xdeada55** [reply]
I wouldn't say it's "very very weak minification", it's just regular minification. 

Anyway, I gotta give you credit on not obfuscating and including source maps, that is interesting. I'm wondering why even bother not wanting to open source it when you are providing sourcemaps. 

You could open source it, make the project gain traction and you get reputation for it, maybe even get your goal of it being sponsored/getting a donation from a company.

**Reply by igor-lemon** [reply]
Hehe, really in 2019 I included source maps for production bundles, I don't remember why I did it, but maybe I had some reason for it. But it's not critical for me, because I have nothing to hide and we didn't make any magic inside.

I wrote that I couldn't do it without approve of all owners.
Have you tried to do some open source projects? I have some experience. If you are not ready to support the project yourself, then no one from the outside will spend time supporting it. Perhaps big and popular projects will be an exception of the rule. But first you have to become a big project. Chicken-and-Egg Problem.

I will soon try to publish an article about DockStation 2.0 plans. 
Because I know some problems which don't solve any product on the market and some issues which I couldn't solve via DockStation easly in my daily tasks at work.
If some company(s) will support me it will be OpenSource project, if not it will be propietary commercial product.

**Reply by jamezrin** [reply]
> Hehe, really in 2019 I included source maps for production bundles, I don't remember why I did it, but maybe I had some reason for it. But it's not critical for me, because I have nothing to hide and we didn't make any magic inside.
> I wrote that I couldn't do it without approve of all owners. Have you tried to do some open source projects? I have some experience. If you are not ready to support the project yourself, then no one from the outside will spend time supporting it. Perhaps big and popular projects will be an exception of the rule. But first you have to become a big project. Chicken-and-Egg Problem.
> 
That is a very good reasoning for making it open source as well, but I understand what you are saying regarding other codeowners. 

I get what you are saying, you are 100% right on that if you can't dedicate time to the project, no one will, it's rare projects get contributions. I know this first hand, I maintain projects that are popular and unless we fix problems no one will (90%+ of the time).
Regardless, I think in this case is more of a trust thing rather than being able to contribute. 

Please, understand that from my point of view this could very well include malware against the west for things like sanctions and everything you are getting in Belarus. I don't know your point of view on the situation, but understand that we are skeptical about Russians and Belorussians in general. No ill intent, just precaution.

> I will soon try to publish an article about DockStation 2.0 plans. Because I know some problems which don't solve any product on the market and some issues which I couldn't solve via DockStation easly in my daily tasks at work. If some company(s) will support me it will be OpenSource project, if not it will be propietary commercial product.

I wish you the best with DockStation 2, I really liked this app 3 years ago and would love to use it again in the future.

**Reply by igor-lemon** [reply]
> Please, understand that from my point of view this could very well include malware against the west for things like sanctions and everything you are getting in Belarus. I don't know your point of view on the situation, but understand that we are skeptical about Russians and Belorussians in general. No ill intent, just precaution.

My mother form Ukraine, I have many relatives in Ukraine and I try to help as much as I can. I wish burn to hell to everyone occupiers who came to Ukraine and make a war there. Please don't hate belarusians today. I propose you you to know what happened in Belarus in 2020. Today Belarus is North Korea in Europe. You can simply be recognized as an extremist and put in jail for years if you just express your opinion. It's not joke, it's sad reallity. More than a thousand political prisoners are absolute world records. Total intimidation and expulsion of best people from the country. It's really sad...

**Reply by jamezrin** [reply]
I'm sorry, I didn't want to express hate on your people, it's the politicians and the ones pro-war that are to blame.
I wish you and yours the best, hopefully you are safe.

**Reply by igor-lemon** [reply]
Thanks! I hope everything goes well 🙏🏻

**Comment by softexpert** [comment]
Thank you for the insight !

In my view, the most important question is how big and motivated is the community around the Dockstation project, to start building from scratch a better version?
Approaches can be as many as there are frameworks; based on previous experience, I would say:
- fully open source (MIT or GPL licensed)
- easy to build
- multi-platform
- web based interface

Personally, I would incline towards the following:
- using some javascript libraries like MithrilJS, Vue.js, etc, for the front-end
- do not use Electron - meaning no weird, heavy and complicated dependencies to maintain
- implementing in GO language the kernel of the app - meaning that all ressources can be embedded in a single executable file and the interface being loaded in any standard browser; also taking advantage of existing projects written in GO that can inspect and analyse existing docker and podman containers

What do you think?

**Reply by igor-lemon** [reply]
1) The community is not very big, but not small, thousands of people use the application every day. I'm sure I can raise the user's count.

2) I'm ready to make Open Source but I understand that ~98% of work will be on my side and I wanna be motivated. Just need to find a few big companies who will be ready to support us. If not, just commercial product.

> easy to build
multi-platform
web based interface

 I agree with you 100%

=====

> using some javascript libraries like MithrilJS, Vue.js, etc, for the front-end

I'm a frontend dev :) Yep, I'll use React because I have a big experience in development with React (the current DockStation is based on React too)

=====

>do not use Electron - meaning no weird, heavy and complicated dependencies to maintain

I agree. But I have plans to use Electron but 
1) it will be such as a helper which can help to install and run everything with 1 click
2) just implement `BrowserView` to show the web interface

=====

>implementing in GO language the kernel of the app - meaning that all ressources can be embedded in a single executable file and the interface being loaded in any standard browser; also taking advantage of existing projects written in GO that can inspect and analyse existing docker and podman containers

About backend. It's an open question. I have experience in NodeJS my friend has an experience in Python. I would prefer to use Python because the number of developers who know Python is much more than Go. However, I agree that using Go might be better because of the better performance and the fact that Docker is written with Go, which allows getting some solutions and code reuse.

**Reply by softexpert** [reply]
I'm guessing a lot of good ideas related to Docker can be spotted in projects like [dive](https://github.com/wagoodman/dive) (MIT license).

As for the GO language, I see the following benefits:
- same code can be compiled for different target platforms
- ressources (javascript libraries and components, CSS, images) can be embedded in the executable
- the result of the compilation is a single file that contains everything - so, very easy to deploy
- it is fast
- starting from go v1.18 generics can be used, which might simplify some of the code architecture

I have built for my needs such an application; all I needed was to copy the exe file (I compiled in Linux targeting Windows) on the server and launch the application; right there I already had a browser (Chrome, Edge, Firefox), so I could instantly have access to the application UI (it can even launch on start the default web browser with the local URL).

Nothing beats that kind of comfort - easy to develop, easy to deploy, easy to use.

Adding node.js in the loop looks nightmarish for me, because there are relentless updates and changes, and the dependencies are way too complex for the final purpose.
Could be used to produce the distributable part for the JS based UI, but I would stop there.

That is my view, ofcourse; everyone will tend to embrace the most familiar approach; I just felt the need to share my experience with you, hoping it could be useful.

**Reply by igor-lemon** [reply]
I can agree with you, but first I need someone who good knows Go, who has big dev experience, who wants to do it project, and who is ready to commit min 1.5-2 years to the project. If you're this dev and you're ready let's connect on LinkedIn and we can discuss it.

**Comment by captnbli** [comment]
I'm working for a company now, and I couldn't recommend using this for now. Apart from the lack of documentation (what are the possible values for the docker "driver"?, there is the issue of community, support and longevity. Open sourcing might unlock all of these issues.
